src/main/java/emu/grasscutter/game/managers/EnergyManager.java

```diff
@@ -0,0 +1,215 @@
+		// In case we can't, we default to an elementless particle.
+		// The element is the element of the avatar that has cast the ability.
+		// We can get that from the avatar's skill depot.
+		int itemId = 2024;
+		
+		// Try to fetch the avatar from the player's party and determine their element.
+		// ToDo: Does this work in co-op?
+		int avatarId = getCastingAvatarIdForElemBall(invoke.getEntityId());	
+		Optional<EntityAvatar> avatarEntity = player.getTeamManager().getActiveTeam()
+													.stream()
+													.filter(character -> character.getId() == avatarId)
+													.findFirst();
+
+		if (avatarEntity.isPresent()) {
+			Avatar avatar = avatarEntity.get().getAvatar();
+
+			if (avatar != null) {
+				AvatarSkillDepotData skillDepotData = avatar.getSkillDepot();
+
+				if (skillDepotData != null) {
+					ElementType element = skillDepotData.getElementType();
+
+					// If we found the element, we use it to deterine the ID of the
+					// energy particle that we have to generate.
+		 			if (element != null) {
+		 				itemId = switch (element) {
+							case Fire -> 2017;
+							case Water -> 2018; 
+							case Grass -> 2019;
+							case Electric -> 2020;
+							case Wind -> 2021;
+							case Ice -> 2022;
+							case Rock -> 2023;
+							default -> 2024;
+						};
+		 			}
+		 		}
+		 	}
+		}
+
+		// Get the item data for an energy particle of the correct element.
+		ItemData itemData = GameData.getItemDataMap().get(itemId);
+		if (itemData == null) {
+			return; // Should never happen
+		}
+		
+		// Generate entity.
+		EntityItem energyBall = new EntityItem(getPlayer().getScene(), getPlayer(), itemData, new Position(action.getPos()), 1);
+		energyBall.getRotation().set(action.getRot());
+
+		this.getPlayer().getScene().addEntity(energyBall);
+	}
+
+    /**********
+        Pickup of elemental particles and orbs.
+    **********/
+    public void handlePickupElemBall(GameItem elemBall) {
+        // Check if the item is indeed an energy particle/orb.
+		if (elemBall.getItemId() < 2001 ||elemBall.getItemId() > 2024) {
+			return;
+		}
+
+		// Determine the base amount of energy given by the particle/orb.
+		// Particles have a base amount of 1.0, and orbs a base amount of 3.0.
+		float baseEnergy = (elemBall.getItemId() <= 2008) ? 3.0f : 1.0f;
+
+		// Add energy to every team member.
+		for (int i = 0; i < this.player.getTeamManager().getActiveTeam().size(); i++) {
+			EntityAvatar entity = this.player.getTeamManager().getActiveTeam().get(i);
+
+			// On-field vs off-field multiplier.
+			// The on-field character gets no penalty.
+			// Off-field characters get a penalty depending on the team size, as follows:
+			//	  - 4 character team: 0.6
+			//	  - 3 character team: 0.7
+			//    - 2 character team: 0.8
+			float offFieldPenalty = (this.player.getTeamManager().getCurrentCharacterIndex() == i) ? 1.0f : 1.0f - this.player.getTeamManager().getActiveTeam().size() * 0.1f;
```

Comment:
Make sure `offFieldPenalty` does not go below 0 (or .1) in case someone has more than 4 avatars in their party.

Reply:
Oh, good point, I didn't consider that. Do you happen to know what the official server does when we have a 5th trial character in the party? Does it continue to scale down, or does it continue doing what it does for a 4 character party?